Remove the boost dependency

svengoldberg opened the pull request:
All boost commands, functions etc. are removed and replaced by `std` code from the C++17-standard. The testsuite was replaced by GTest which will be downloaded from GitHub and configured within CMake.

As the test framework is a critial part of a software, it was specifically tested locally. Here, e.g., errors were provoked which showed the wanted results.


Fixes #3

Additionally, the minimum required CMake version is raised to 3.11 to avoid potential future deprecation warnings. This step also reduces code in the GTest-CMake-file because no differentiation between version number has to be made.


Fixes #5

svengoldberg committed:
Remove the boost dependency

All boost commands, functions etc. are removed and replaced by std code from CXX17. The testsuite was replaced by GTest which is copied into the repository and installed locally

svengoldberg committed:
Edit cmake.yml

Remove boost dependency and make log output verbose

svengoldberg committed:
Include <functional> in CodeGen.cpp

svengoldberg committed:
Remove hard-wired gtest source code, download from GitHub

joergbrech commented on README.md:
```diff
@@ -42,9 +41,7 @@ For each subdirectory, CPACSGen runs an additional pass, putting the generated f
 2. Clone the TiGL repository Github
 3. Use CMake to configure CPACSGen
-    1. Set BOOST_ROOT
-    2. Set TIXI_PATH
```
We should keep this line

svengoldberg replied:
Ok, I see the point... To me, one would just build the generator within the `tigl-bld`-env, anyway. So, `tixi` is found without further specification. Maybe, we could write:

> 1. Set TIXI_PATH, if the TiXI is not already installed in the conda environment

joergbrech commented on .github/workflows/cmake.yml:
```diff
@@ -24,16 +24,17 @@ jobs:
+      working-directory: ${{github.workspace}}/build
       # Build your program with the given configuration
-      run: cmake --build ${{github.workspace}}/build --config ${{env.BUILD_TYPE}}
+      run: make VERBOSE=1
```
Why was this changed?

svengoldberg replied:
I did this during fixing the CI to understand the error message. I found it reasonable to see the exact compile commands for potential future issue fixes. That is why, I wanted to have verbose output

joergbrech replied:
Ok, I guess its fine, since we are only testing on Linux so far. I just wanted to understand the changes and if anything was wrong with the original implementation.

joergbrech commented on src/lib/CodeGen.cpp:
```diff
@@ -1295,15 +1334,14 @@ namespace tigl {
-                                            const auto it = std::unique(std::begin(v), std::end(v));
-                                            v.erase(it, std::end(v));
+                                            std::sort(v.begin(), v.end());
+                                            v.erase(std::unique(v.begin(), v.end()), v.end());
```
This lambda `unique` is duplicated. Maybe we should promote it to a function in the outer scope and reuse it?

BTW: Why did you change it? The original code seemed fine to me.

svengoldberg replied:
Oh yeah thanks, I did not see this... I will define the function `unique` within the `CodeGen` class.

joergbrech commented on src/lib/CodeGen.cpp:
```diff
@@ -1295,15 +1334,14 @@ namespace tigl {
 
-                                        const auto it = std::remove_if(std::begin(allIndices), std::end(allIndices), [&](std::size_t ai) {
-                                            for (const auto ci : childIndices) {
+                                        auto it = std::remove_if(allIndices.begin(), allIndices.end(), [&](std::size_t ai) {
+                                            for (std::size_t ci : childIndices) {
```
Why did you change this code? Not a big deal, but I must admit that I prefered the original code, especially due to the `const` specifiers. 

I don't actually mind the changes from `std::begin(x)` to `x.begin()`, but strictly speaking `std::begin` and `std::end` are more general, because they accept some types that don't have `begin` and `end` methods.

svengoldberg replied:
I will change this back to the old version!

joergbrech commented on src/lib/Tables.cpp:
```diff
@@ -75,15 +75,15 @@ namespace tigl {
 
-    boost::optional<const std::string&> MappingTable::find(const std::string& key) const {
+    const std::string* MappingTable::find(const std::string& key) const {
         const auto it = m_map.find(key);
         if (it != std::end(m_map))
-            return it->second;
+            return &(it->second);
         else
-            return {};
+            return nullptr;
     }
```
I liked the version with an optional return type better. If you  changed this because `std::optional` doesn't allow references, you could return a `std::optional<std::string_view>`: https://godbolt.org/z/xK3n3Tj9M

svengoldberg replied:
Yes, that was the exact reason! The next hurdle here would be that the return type then changes from a `std::string` reference to a `std::string_view` (when it has a value). That needs extra treatment at every occurence (which are quite a lot) when indeed a `std::string` is expected. Or could you think of a wrapper function around `MappingTable::find(...)` which would be called and translate to `std::string`?
But it would also just translate the problem elsewhere... Or maybe you have a better solution here!

joergbrech replied:
Have you tried it? You should be able to explicitly convert an `std::string_view` to a `std::string`, right? See ctor no 9 [here](https://en.cppreference.com/w/cpp/string/basic_string/basic_string.html).

https://godbolt.org/z/4bYoeaKWd

joergbrech replied:
Another alternative would be `std::reference_wrapper`, I just thought this would be simpler. Now when having to choose between your pointer implementation and `std::reference_wrapper` I couldn't say which I prefer, both have their pros and cons. 

BTW, optional references are planned for c++26, not sure we want to sit this out tho... ;)

svengoldberg replied:
Info: We found a solution outside from GitHub

joergbrech commented on src/lib/TypeSystem.cpp:
```diff
@@ -76,9 +79,9 @@ namespace tigl {
+                struct ContentVisitor {
                     ContentVisitor(const xsd::SchemaTypes& types, std::vector<Field>& members, ChoiceElements& choiceItems, std::size_t attributeCount, const Tables& tables, std::vector<std::size_t> choiceIndices = {})
-                        : types(types), members(members), choiceItems(choiceItems), attributeCount(attributeCount), tables(tables), choiceIndices(choiceIndices) {}
+                        : types(types), members(members), choiceItems(choiceItems), attributeCount(attributeCount), tables(tables), choiceIndices(std::move(choiceIndices)) {}
```
Why the `std::move` here? It smells a bit to me, because the argument is passed in by value and then moved into the member. Why is it passed in by value in the first place? (I have to admit I haven't checked how this function is used in code).

svengoldberg replied:
I would say, that the parameter is passed by value because it has a default value. Or is it possible to pass a value by reference when it should have a default value?

Regarding the `move`: You are right, I will reMOVE it!

joergbrech replied:
Ah of course! I suppose its fine then. I guess the move actually makes sense also. I was just not paying attention.

joergbrech commented on src/lib/runtime/TixiHelper.h:
```diff
@@ -19,14 +19,20 @@
-#ifndef BOOST_DATE_TIME_NO_LIB
-    #define BOOST_DATE_TIME_NO_LIB
+#ifndef DATE_TIME_NO_LIB
+    #define DATE_TIME_NO_LIB
```
Isn't this BOOST specific? Do we need it?

svengoldberg replied:
Tbh, I cannot really tell. But indeed, it sounds reasonable to me that is it not needed.

joergbrech commented on test/main.cpp:
```diff
@@ -1,5 +1,5 @@
-#define BOOST_TEST_MODULE CPACSGenTests
-#include <boost/test/unit_test.hpp>
+#define TEST_MODULE CPACSGenTests
```
Isn't this BOOST specific? Can it be removed?

svengoldberg replied:
Same as above...
I will try without both macros

joergbrech commented on test/main.cpp:
```diff
@@ -27,60 +30,61 @@ void runTest() {
 
-BOOST_AUTO_TEST_CASE(sequence) {
+TEST(CPACSGenTests, sequence) {
     runTest();
```
Are all tests calling the same function? This doesn't seem right. Also, if I look at the output of the unit tests in CI I see that the output of all tests is equal and it is actually a failure: The tests fail to read some input files. So there are some To Dos for the unit tests:

 - Make sure all test files are found
 - Make sure the tests are parametrized properly
 - Make sure that the return code of the test executable is not zero if a a test fails. Otherwise we get a green checkmark in CI

svengoldberg replied:
Yes, they do all call the same function.
Inside the `runTest()` function, gtest-functions are called to return the test name. All the test names listed in this `test/main.cpp` also appear as directories in `test/data/`-
The `runTest()` function then generates different files and compares them each with different reference files located at `test/data/<testName>/ref.cpp`. I actually liked the "clean" look that has no repitition. One could think of passing a parameter to the `runTest()` function which is this very directory name to make clear what happens under the hood without having to look into the function code.

When I tested it, it did not get the same output in all tests. Also, when I manipulated a reference file (to provoke an error when the comparison breaks and see whether the test fails) only this very test fails.

Regarding the test files: Some of them really just do not exist in the repository. I did not have a deep-dive look into the actual code of the reference files... But I assumed that the respective missing files are just not relevant in the corresponding test. Those files also appear in the `tigl` directory `cpacs_gen_input/`. They seem to be needed for the complete and actual file generation for TiGL but not for the unit tests, as far as I can tell. For that reason, I would say that this is not an actual fail. Indeed, when a test file is not generated correctly, the tests do fail.
Nevertheless, you are right: It is indeed a little ugly that the test output is that way.

svengoldberg committed:
Remove relicts from boost package

svengoldberg committed:
Remove warning in test due to missing namespace in .xsd

svengoldberg committed:
Empty dummy files to avoid warnings in tests

Up to now, most of the tests throw 'Warnings' because files cannot be found.
Indeed, those files are not needed by the respective unit test itself.
But nevertheless, in production they are required when generating TiGL code (see tigl/cpacs_gen_input/).
Hence, the code prints a warning. To avoid this, empty files with the required name are included.

svengoldberg committed:
Edit README.md

svengoldberg committed:
Remove absolete .yml files from old CI

svengoldberg committed:
Change type of MappingTable::find() to std::optional<...>

svengoldberg committed:
Make code DRY by rewriting lambda to function

joergbrech commented on README.md:
```diff
@@ -15,7 +15,6 @@ The following components are required to build and run CPACSGen
 * CMake
 * C\++11/C++14 compliant compiler (VS2017 is currently used and tested)
```
We actually need a C++17 compliant compiler now.

svengoldberg replied:
Of course... You are right, thanks!

svengoldberg committed:
Edit README.md (compiler for C++17 needed)

svengoldberg committed:
Replace v.end() by std::end()

svengoldberg committed:
Remove std::move() in constructor

svengoldberg committed:
Reactivate std::move() in constructor

joergbrech reviewed (approved):
Nice! Looks good to me. I haven't tested it locally yet, but the code changes seem reasonable. Thanks!
